barretenberg/cpp/pil/avm/avm_alu.pil

```diff
@@ -474,3 +476,75 @@ namespace avm_alu(256);
     (p_sub_b_lo' - res_lo) * rng_chk_sel'= 0;
     (p_sub_b_hi' - res_hi) * rng_chk_sel'= 0;
 
+    // ========= SHIFT LEFT/RIGHT OPERATIONS ===============================
+    // Given inputs to a shift operation, a & b, and a memory tag, mem_tag.
+    // Split a into Big Endian hi and lo limbs, a_hi and a_lo, and the number of bits represented by the memory tag, t.
+    // QUESTION: SHOULD B BE CONSTRAINED TO BE U8 -> i.e. when would we shift more than 255 bits when the max number of bits of a is 128bits?
+    //
+    // === IF SHR
+    // (1) Prove the correct decomposition: a_hi * 2**b + a_lo = a
+    // (2) Range check a_hi < 2**(t-b) && a_lo < 2**b, ensure we have not overflowed the limbs during decomp
+    // (3) Return a_hi
+    //
+    //   <--(t-b) --> | Length: b
+    // ---------------|-----------------------
+    // |      a_hi    |          a_lo        | --> a
+    // ---------------------------------------
+    //
+    // === IF SHL
+    // (1) Prove the correct decomposition: a_hi * 2**(t-b) + a_lo = a
+    // (2) Range check a_hi < 2**b && a_lo < 2**(t-b)
+    // (3) Return a_lo * 2**b
+    // 
+    //          <-- b-->      |   <-- (t-b) -->
+    // -----------------------|---------------
+    // |      a_hi            |  a_lo        | --> a
+    // ---------------------------------------
+
+    // TODO: Variable Range Checks
+    // Check that a_lo and a_hi combine to be the same number of bits represented by the Memory Tag
+    // Can we can convert this into a fixed range check over b that might be more optimal for us? 
+    // SHR: a_hi < 2**(t - b) && a_lo < 2**b
```

Comment:
For SHR, it is sufficient to prove that a_hi * 2**b does not overflow the field order. Since, we can assume b < t, we can round up the range check for a_hi to the next multiple of 16 bits or in the worst case showing <2^t
However, we do need the range check for a_lo to be exactly within b bits, this can be achieved by showing that 2**b - a_lo < 2^t or whatever power of two lower than 2^t.

Reply:
Agreed - Keeping the notes we discussed before here for my reference
```
0 < 2**(t - b) - a_hi < 2^128
0 < 2**b - a_lo < 2^128

2^b * a_hi + a_lo = a
````